pkg/tar/chroot.go

```diff
@@ -0,0 +1,107 @@
+// Copyright 2015 The rkt Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tar
+
+import (
+	"archive/tar"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"syscall"
+
+	"github.com/coreos/rkt/pkg/multicall"
+)
+
+const (
+	multicallName = "extracttar"
+	fileMapFdNum  = 3
+)
+
+var mcEntrypoint multicall.Entrypoint
+
+func init() {
+	mcEntrypoint = multicall.Add(multicallName, extractTarCommand)
+}
+
+func extractTarCommand() error {
+	if len(os.Args) < 3 {
+		return fmt.Errorf("not enough arguments")
+	}
+	dir := os.Args[1]
+	if !filepath.IsAbs(dir) {
+		return fmt.Errorf("dir %s must be an absolute path", dir)
+	}
+	overwrite, err := strconv.ParseBool(os.Args[2])
+	if err != nil {
+		return fmt.Errorf("error parsing overwrite argument: %v", err)
+	}
+
+	if err := syscall.Chroot(dir); err != nil {
+		return fmt.Errorf("failed to chroot in %s: %v", dir, err)
+	}
+	if err := syscall.Chdir("/"); err != nil {
+		return fmt.Errorf("failed to chdir: %v", err)
+	}
+	fileMapFile := os.NewFile(uintptr(fileMapFdNum), "fileMap")
+
+	fileMap := map[string]struct{}{}
+	if err := json.NewDecoder(fileMapFile).Decode(&fileMap); err != nil {
+		return fmt.Errorf("error decoding fileMap: %v", err)
+	}
+	if err := extractTar(tar.NewReader(os.Stdin), overwrite, fileMap); err != nil {
+		return fmt.Errorf("error extracting tar: %v", err)
+	}
+
+	// flush remaining bytes
+	io.Copy(ioutil.Discard, os.Stdin)
+
+	return nil
+}
+
+// ExtractTar extracts a tarball (from a io.Reader) into the given directory
```

Comment:
two things:
1) please add a note that this will chroot and hence requires privileges
2) should we add a check here for that priv? maybe that's slightly more elegant than failing during the `Chroot()`? I'm on the fence